src/dbnode/storage/index.go

```diff
@@ -1594,89 +1625,62 @@ func (i *nsIndex) queryWithSpan(
 			break
 		}
 
-		if applyTimeout := timeout > 0; !applyTimeout {
+		wg.Add(1)
```

Comment:
this could use some close eyes. it should behave exactly as before. instead of launching a goroutine to watch the timeout and return, we are using a goroutine to watch the timeout and interrupt. when the workers are interrupted, it will ultimately return preserving the existing behavior and properly cleaning up.

Reply:
Yeah I think this looks good.